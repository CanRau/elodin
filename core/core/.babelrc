{
  "plugins": ["@babel/plugin-transform-async-to-generator"],
  "presets": [
    [
      "@babel/preset-env",
      {
        "modules": false
      }
    ]
  ],
  "env": {
    "commonjs": {
      "plugins": ["transform-es2015-modules-commonjs"]
    }
  }
}
