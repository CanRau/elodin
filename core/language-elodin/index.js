var vscode = require('vscode')

var properties = [
  'alignContent',
  'alignItems',
  'alignSelf',
  'backgroundColor',
  'borderBottomColor',
  'borderBottomLeftRadius',
  'borderBottomRightRadius',
  'borderBottomWidth',
  'borderColor',
  'borderLeftColor',
  'borderLeftWidth',
  'borderRadius',
  'borderRightColor',
  'borderRightWidth',
  'borderStyle',
  'borderTopColor',
  'borderTopLeftRadius',
  'borderTopRightRadius',
  'borderTopWidth',
  'borderWidth',
  'bottom',
  'color',
  'direction',
  'display',
  'flexBasis',
  'flexDirection',
  'flexGrow',
  'flexShrink',
  'flexWrap',
  'fontFamily',
  'fontSize',
  'fontStyle',
  'fontWeight',
  'height',
  'justifyContent',
  'left',
  'letterSpacing',
  'lineHeight',
  'margin',
  'marginBottom',
  'marginLeft',
  'marginRight',
  'marginTop',
  'maxHeight',
  'maxWidth',
  'minHeight',
  'minWidth',
  'opacity',
  'padding',
  'paddingBottom',
  'paddingLeft',
  'paddingRight',
  'paddingTop',
  'position',
  'right',
  'textAlign',
  'textDecorationLine',
  'top',
  'width',
]

var valueMap = {
  direction: ['ltr', 'rtl'],
  position: ['relative', 'absolute'],
  flexBasis: ['auto'],
  borderStyle: ['solid', 'dotted', 'dashed'],
  flexWrap: ['nowrap', 'wrap'],
  flexDirection: ['row', 'rowReverse', 'column', 'columnReverse'],
  justifyContent: [
    'flexStart',
    'flexEnd',
    'center',
    'spaceBetween',
    'spaceAround',
    'spaceEvenly',
  ],
  alignSelf: ['auto', 'flexStart', 'flexEnd', 'center', 'stretch', 'baseline'],
  alignItems: ['flexStart', 'flexEnd', 'center', 'stretch', 'baseline'],
  alignContent: [
    'flexStart',
    'flexEnd',
    'center',
    'stretch',
    'spaceBetween',
    'spaceAround',
  ],
  fontStyle: ['normal', 'italic'],
  textAlign: ['auto', 'justify', 'center', 'left', 'right'],
  textDecorationLine: ['none', 'underline', 'lineThrough'],
}

const values = [
  'absolute',
  'auto',
  'baseline',
  'center',
  'column',
  'columnReverse',
  'dashed',
  'dotted',
  'flexEnd',
  'flexStart',
  'hide',
  'italic',
  'justify',
  'left',
  'lineThrough',
  'ltr',
  'none',
  'normal',
  'nowrap',
  'relative',
  'right',
  'row',
  'rowReverse',
  'rtl',
  'show',
  'solid',
  'spaceAround',
  'spaceBetween',
  'spaceEvenly',
  'stretch',
  'underline',
  'wrap',
]
var functions = ['rgb', 'hsl', 'rgba', 'hsla', 'hex', 'percentage']
var keywords = ['view', 'text', 'variant', 'fragment']
var mediaQueries = ['viewportWidth', 'viewportHeight', 'landscape', 'portrait']
var pseudoClasses = [
  'active',
  'checked',
  'default',
  'disabled',
  'empty',
  'enabled',
  'first',
  'firstChild',
  'firstOfType',
  'focus',
  'focusWithin',
  'hover',
  'inRange',
  'intermediate',
  'invalid',
  'lastChild',
  'lastOfType',
  'left',
  'link',
  'onlyChild',
  'onlyOfType',
  'optional',
  'readOnly',
  'readWrite',
  'required',
  'right',
  'target',
  'valid',
  'visited',
]
var pseudoElements = [
  'after',
  'before',
  'firstLetter',
  'firstLine',
  'selection',
]

vscode.languages.registerCompletionItemProvider(
  { language: 'elodin' },
  {
    provideCompletionItems(document, position, token, context) {
      if (context.triggerCharacter !== ':') {
        return []
      }

      var textBefore = document
        .lineAt(position)
        .text.substr(0, position.character - 1)
      var splitted = textBefore.split(' ').pop()

      if (valueMap[splitted]) {
        return valueMap[splitted].map(function(p) {
          return new vscode.CompletionItem(p)
        })
      }
    },
  },
  ':'
)

vscode.languages.registerCompletionItemProvider(
  { language: 'elodin' },
  {
    provideCompletionItems(document, position, token, context) {
      if (context.triggerCharacter !== '@') {
        return []
      }
      return [...mediaQueries, ...pseudoClasses, ...pseudoElements].map(
        function(p) {
          return new vscode.CompletionItem(p)
        }
      )
    },
  },
  '@'
)

vscode.languages.registerCompletionItemProvider(
  { language: 'elodin' },
  {
    provideCompletionItems(document, position, token, context) {
      if (
        context.triggerCharacter === ':' ||
        context.triggerCharacter === '@'
      ) {
        return []
      }

      const textToChar = document
        .lineAt(position)
        .text.substr(0, position.character)
      var textBefore = textToChar.trim()

      if (textBefore.charAt(textBefore.length - 1) === ':') {
        var splitted = textBefore
          .substr(0, textBefore.length - 1)
          .split(' ')
          .pop()

        if (valueMap[splitted]) {
          return valueMap[splitted].map(function(p) {
            return new vscode.CompletionItem(p)
          })
        }
      }

      return [
        ...properties,
        ...keywords,
        ...values,
        ...mediaQueries,
        ...pseudoClasses,
        ...pseudoElements,
        ...functions,
      ].map(function(p) {
        return new vscode.CompletionItem(p)
      })
    },
  }
)
