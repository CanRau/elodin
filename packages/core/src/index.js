import { parse, errorTypes } from '@elodin/parser'
import { traverse } from '@elodin/traverser'
import { format, formatFromAST } from '@elodin/format'
import * as types from '@elodin/types'

import transformFile from './transformFile'

export {
  parse,
  traverse,
  format,
  formatFromAST,
  types,
  transformFile,
  errorTypes,
}
