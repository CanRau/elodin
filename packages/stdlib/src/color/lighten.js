import Color from 'color'

export default function lighten(color, percentage, type = 'rgb') {
  return Color(color)
    .lighten(percentage)
    [type]()
    .string()
}
