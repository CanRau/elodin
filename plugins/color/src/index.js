const defaultSelector = (variables, property) => variables[property]

// available formats: rgb, hsl, hex, keyword
const defaultConfig = {
  format: 'rgb',
}

export default function color(customConfig = {}) {
  const { format } = {
    ...defaultConfig,
    ...customConfig,
  }

  return {
    Color: {
      enter(path) {
        path.node.format = format
      },
    },
  }
}
