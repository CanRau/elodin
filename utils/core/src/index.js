import getVariablesFromAST from './getVariablesFromAST'
import getVariantsFromAST from './getVariantsFromAST'

export { getVariablesFromAST, getVariantsFromAST }
