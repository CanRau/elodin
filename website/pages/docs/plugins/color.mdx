import { make as DocLayout } from '../../../components/DocLayout.bs.js'

export default DocLayout

# Color
