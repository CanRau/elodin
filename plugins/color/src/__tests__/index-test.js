import { traverse, parse } from '@elodin/core'

import color from '../index'

describe('Normalizing colors', () => {
  it('should correctly set color formats', () => {
    const file = `view Button { backgroundColor: red borderBottomColor: rgb(255 100 50) }`

    const ast = parse(file).ast

    expect(traverse(ast, [color()])).toMatchSnapshot()
  })
})
