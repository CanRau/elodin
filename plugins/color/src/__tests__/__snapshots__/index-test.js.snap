// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Normalizing colors should correctly set color formats 1`] = `
Object {
  "body": Array [
    Object {
      "body": Array [
        Object {
          "dynamic": false,
          "property": "backgroundColor",
          "raw": false,
          "type": "Declaration",
          "value": Object {
            "alpha": 1,
            "blue": 0,
            "format": "hex",
            "green": 0,
            "red": 255,
            "type": "Color",
          },
        },
        Object {
          "dynamic": false,
          "property": "borderBottomColor",
          "raw": false,
          "type": "Declaration",
          "value": Object {
            "alpha": 1,
            "blue": 50,
            "format": "hex",
            "green": 100,
            "red": 255,
            "type": "Color",
          },
        },
      ],
      "format": "view",
      "name": "Button",
      "type": "Style",
    },
  ],
  "type": "File",
}
`;
